### PR TITLE
ref(notifications): switch to a role based strategy

### PR DESCRIPTION
**Pull request by scefali**
This PR adds the concepts of a strategy class that we can pass to notifications to add functionality to certain types of notifications without a mix-in. I've added a `RoleBasedStrategy` class to handle functionality when we need to target organization members based on a specific role.

Getsentry PR: https://github.com/getsentry/getsentry/pull/6773

### CODE REVIEW

File: src/sentry/notifications/notifications/strategies/role_based_strategy.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from typing import TYPE_CHECKING, Iterable, MutableMapping
+
+from sentry import roles
+from sentry.models import OrganizationMember, Team
+
+if TYPE_CHECKING:
+    from sentry.models import Organization, User
+
+
+class RoleBasedStrategy:
```
Comment (JoshFerge) [inline src/sentry/notifications/notifications/strategies/role_based_strategy.py:12]: is this an ABC?

Comment (scefali) [reply src/sentry/notifications/notifications/strategies/role_based_strategy.py:12]: Yep, guess I should inherit from it

File: src/sentry/notifications/notifications/strategies/role_based_strategy.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        raise NotImplementedError
+
+    def get_role_string(self, member: OrganizationMember) -> str:
+        role_string: str = roles.get(member.role).name
+        return role_string
+
+    def build_notification_footer_from_settings_url(
+        self, settings_url: str, recipient: Team | User
+    ) -> str:
+        if isinstance(recipient, Team):
```
Comment (JoshFerge) [inline src/sentry/notifications/notifications/strategies/role_based_strategy.py:56]: I would move this type check to the caller and type recipient as `User` here, but your call, maybe a `TeamRecipientStrategy` could clean this param type later on

File: src/sentry/notifications/notifications/strategies/role_based_strategy.py (hunk 1 above)
Comment (JoshFerge) [inline src/sentry/notifications/notifications/strategies/role_based_strategy.py:12]: only other nit, would call this `RoleBasedRecipientStrategy`

**Review by JoshFerge — APPROVED** [review]

File: src/sentry/notifications/notifications/strategies/role_based_strategy.py (hunk 1 above)
Comment (JoshFerge) [reply src/sentry/notifications/notifications/strategies/role_based_strategy.py:12]: you can use the `@abstractmethod` decorator below if you do that.

**Pull request merged**